pennylane/resource/error.py

```diff
@@ -77,3 +78,75 @@ def error(self) -> AlgorithmicError:
         Returns:
             AlgorithmicError: The error.
         """
+
+
+class SpectralNormError(AlgorithmicError):
+    """Class representing the spectral norm error.
+
+    The spectral norm error is defined as the distance, in spectral norm, between the true unitary
+    we intend to apply and the approximate unitary that is actually applied.
+
+    Args:
+        error (float): The numerical value of the error
+
+    **Example**
+
+    >>> s1 = SpectralNormError(0.01)
+    >>> s2 = SpectralNormError(0.02)
+    >>> s3 = s1.combine(s2)
+    >>> print(type(s3), s3.error)
+    <class 'pennylane.resource.error.SpectralNormError'> 0.03
+    """
+
+    def __repr__(self):
+        """Return formal string representation."""
+
+        return f"<SpectralNormError({self.error})>"
+
+    def combine(self, other: "SpectralNormError"):
+        """Combine two spectral norm errors.
+
+        Args:
+            other (SpectralNormError): The other instance of error being combined.
+
+        Returns:
+            SpectralNormError: The total error after combination.
+
+        **Example**
+
+        >>> s1 = SpectralNormError(0.01)
+        >>> s2 = SpectralNormError(0.02)
+        >>> s3 = s1.combine(s2)
+        >>> print(s3)
+        <SpectralNormError(0.03)>
+        >>> print(s3.error)
+        0.03
+        """
+        return self.__class__(self.error + other.error)
+
+    @staticmethod
+    def get_error(approximate_op: Operator, exact_op: Operator, **kwargs):
+        """Compute spectral norm error between two operators.
+
+        Args:
+            approximate_op (Operator): The approximate operator.
+            exact_op (Operator): The exact operator.
+
+        Returns:
+            float: The error between the exact operator and its
+            approximation.
+
+        **Example**
+
+        >>> Op1 = qml.RY(0.40, 0)
+        >>> Op2 = qml.RY(0.41, 0)
+        >>> SpectralNormError.get_error(Op1, Op2)
+        0.004999994791668309
+        """
+        if not (exact_op.has_matrix and approximate_op.has_matrix):
+            raise ValueError("The input operator must have a matrix.")
```

Comment:
Not sure we need this, since even if the operator does not have a matrix, we can call `qml.matrix(op)` and it will use the decomposition to compute the matrix.

Reply:
It helps in cases like this, taken from the tests:

```
approx_op = Operation
exact_op = qml.RX(0.1, 1)

SpectralNormError.get_error(approx_op, exact_op)
```
I assume @AmintorDusko added this to catch custom operators that inherit from Operation and have no matrix and decomposition defined. e.g.:

```
class MyOp(Operation):
    ...

SpectralNormError.get_error(MyOp, qml.RX(0.1, 1))
```

Reply:
Yes, indeed.

Reply:
But in the case where neither of them are defined, the decomposition method itself will raise an error. So there isn't a need to make this one. 

I don't have a strong opinion for this, feel free to keep it. Just make sure that if a template has a decomposition, then this error is not raised (i.e we can alternatively use the decomposition when it is given instead of the matrix).

Reply:
> But in the case where neither of them are defined, the decomposition method itself will raise an error. So there isn't a need to make this one.

The following code executes with no error:

```
class MyOp(Operation):
    ...
op = MyOp(0)
```

We can check for `compute_decomposition` and `decomposition` though.

Reply:
@Jaybsoni removed the check and updated the test.